File: defacto-core/src/main/java/org/aksw/defacto/ml/feature/fact/impl/SubjectObjectSimilarityFeature.java
/**
 * 
 */
package org.aksw.defacto.ml.feature.fact.impl;

import org.aksw.defacto.evidence.ComplexProof;
import org.aksw.defacto.evidence.Evidence;
import org.aksw.defacto.ml.feature.fact.AbstractFactFeatures;
import org.aksw.defacto.ml.feature.fact.FactFeature;

import uk.ac.shef.wit.simmetrics.similaritymetrics.Levenshtein;

/**
 * @author Daniel Gerber <dgerber@informatik.uni-leipzig.de>
 *
 */
public class SubjectObjectSimilarityFeature implements FactFeature {
	
    Levenshtein lev		= new Levenshtein();

    @Override
    public void extractFeature(ComplexProof proof, Evidence evidence) {
    	
    	String subjectLabel = evidence.getModel().getSubjectLabel(proof.getLanguage());
    	String objectLabel = evidence.getModel().getObjectLabel(proof.getLanguage());
    	
        proof.getFeatures().setValue(AbstractFactFeatures.SUBJECT_SIMILARITY, 
        		Math.max(lev.getSimilarity(proof.getSubject(), subjectLabel), lev.getSimilarity(proof.getSubject(), objectLabel)));
        proof.getFeatures().setValue(AbstractFactFeatures.OBJECT_SIMILARITY, 
        		Math.max(lev.getSimilarity(proof.getObject(), objectLabel), lev.getSimilarity(proof.getObject(), subjectLabel)));
    }
}

File: python/old/triple/fact_extraction/SubjectObjectSimilarityFeature.py
# -*- coding: utf-8 -*-
"""
@author: JohannesLoevenich
"""
import ComplexProof as cp
import editdistance as dist
class SubjectObjectSimilarityFeature:
    SubjectSimilarity = 0
    ObjectSimilarity = 0
    proof = cp.ComplexProof()
    def __init__(self,proof):
        self.proof = proof
        self.model = proof.get_model()
    def subject_levenstein_similarity(self):
        lang = self.proof.get_language()
        modelSub = self.model.get_subject_label(lang)
        modelObj = self.model.get_object_label(lang)
        proofSub = self.proof.get_subject()
        if dist.levenstein(modelSub,proofSub) > dist.levenstein(modelObj,proofSub):
            self.subject_levenstein_similarity = dist.levenstein(modelSub,proofSub)
        else:
            self.subject_levenstein_similarity = dist.levenstein(modelObj,proofSub)
        
    def object_levenstein_similarity(self,lang):
        lang = self.proof.get_language()
        modelSub = self.model.get_subject_label(lang)
        modelObj = self.model.get_object_label(lang)
        proofObj = self.proof.get_object()
        if dist.levenstein(modelSub,proofObj) > dist.levenstein(modelObj,proofObj):
            self.object_levenstein_similarity = dist.levenstein(modelSub,proofObj)
        else:
            self.object_levenstein_similarity = dist.levenstein(modelObj,proofObj)